### ISSUE TITLE
Tune PostgreSQL

### ISSUE THREAD

**Issue by jpmckinney** [open-contracting/deploy#181]
https://wiki.postgresql.org/wiki/Tuning_Your_PostgreSQL_Server

* [x] I see `shared_buffers` is already changed. Others to check: `effective_cache_size`, `work_mem`.  https://wiki.postgresql.org/wiki/Slow_Query_Questions

**Comment by jpmckinney** [comment]
Can consider connection pooling with pgBouncer if needed.

**Comment by jpmckinney** [comment]
@RobHooper 

> [WARN]    checkpoint_completion_target (0.5) is low
> [INFO]    Given those settings PostgreSQL may (depending on its workload) ask the kernel to write (to the storage) up to 1024.00 MB in a timeframe lasting 150 seconds <=> 6.83 MB bytes/second during this timeframe.  You may want to check that your storage is able to cope with this, along with all other I/O (non-writing queries, other software...) operations potentially active during this timeframe.  If this seems inadequate check max_wal_size, checkpoint_timeout and checkpoint_completion_target

`checkpoint_timeout` is the default 5min, which is probably too low. [This post](https://www.2ndquadrant.com/en/blog/basics-of-tuning-checkpoints/) says 30min to 1h is fairly common, in which case `checkpoint_completion_target ` can be 0.9.

In this blog post about another issue, `checkpoint_timeout` is 15min (900) and `checkpoint_completion_target ` is 0.9: https://www.percona.com/blog/2020/07/28/evaluating-checkpointing-in-postgresql/

What do you think?

**Comment by jpmckinney** [comment]
I'll set log_min_duration_statement to 10s, which I think is fairly conservative. We can raise or lower it once we've collected some log messages.

**Comment by jpmckinney** [comment]
With recent commits, we're up from 2-15% memory to 25% on both servers, so tuning definitely makes a difference.

**Comment by jpmckinney** [comment]
Some configurations seems to have come from nowhere in https://github.com/open-contracting/deploy/blob/14fa6ad2191508b745e091aa8d38f0b45c0e489f/salt/postgres/ocdskingfisher_postgres.conf, so I'm restoring them to default settings:

* [max_prepared_transactions](https://www.postgresql.org/docs/current/runtime-config-resource.html#GUC-MAX-PREPARED-TRANSACTIONS) = 100 (default 0) ("If you are not planning to use prepared transactions, this parameter should be set to zero to prevent accidental creation of prepared transactions.)
* [bgwriter_delay](https://www.postgresql.org/docs/current/runtime-config-resource.html#GUC-BGWRITER-DELAY) = 100ms (default 200ms)
* [temp_buffers](https://www.postgresql.org/docs/current/runtime-config-resource.html#GUC-TEMP-BUFFERS) = 128MB (default 8MB)
* [effective_io_concurrency](https://www.postgresql.org/docs/13/runtime-config-query.html#GUC-EFFECTIVE-CACHE-SIZE) = 1000 (1000 is the maximum... now set to 200)

**Comment by jpmckinney** [comment]
To do:

* [ ] checkpoint_completion_target (0.9?) / checkpoint_timeout
* [ ] https://pgbadger.darold.net once we have some logs
* [ ] https://github.com/jfcoz/postgresqltuner again once the system has run under the new settings for a while

**Comment by jpmckinney** [comment]
Some weird things I wasn't expecting:

> FATAL: hot standby is not possible because max_prepared_transactions = 0 is a lower setting than on the master server (its value was 100)
> FATAL: hot standby is not possible because max_worker_processes = 12 is a lower setting than on the master server  (its value was 16)

I don't know if there's a way to change these settings on the replica. For now, I configured them to keep the replica happy.

**Comment by jpmckinney** [comment]
Based on postgresqltuner, we can still increase shared_buffers and work_mem ("PostgreSQL will not use more than 60% of the amount of RAM"). The formula I was following for work_mem might have been too conservative.

**Comment by jpmckinney** [comment]
Now that pg_stat_statements, is enabled, we can also use it to find queries that occupy a lot of time, e.g. https://www.cybertec-postgresql.com/en/pg_stat_statements-the-way-i-like-it/

**Comment by jpmckinney** [comment]
I installed pgbadger and ran `pgbadger /var/log/postgresql/postgresql-11-main.log` from the root user's home directory. I then copied out.html to my local machine to view it. The "Top" and "Events" menus were somewhat interesting.

There are a few ideas here for further tuning (mostly reported by postgresqltuner), but (1) the query load is not very high, so further tuning might not yield much benefit and our logs are not rich enough to derive insights from, and (2) our applications are fairly well optimized on their end (no N+1 queries, etc.), so again the logs might not yield much insight, since we're already writing queries responsibly.
